crates/rpc/rpc-types/src/eth/trace/filter.rs

```diff
@@ -63,27 +63,39 @@ pub struct TraceFilterMatcher {
 impl TraceFilterMatcher {
     /// Returns `true` if the given `from` and `to` addresses match this filter.
     pub fn matches(&self, from: Address, to: Option<Address>) -> bool {
-        // If `from_addresses` and `to_addresses` are empty, then match all transactions.
-        if self.from_addresses.is_empty() && self.to_addresses.is_empty() {
-            return true;
+        let from_matches = self.from_addresses.contains(&from);
+        let to_matches = to.map_or(false, |to_addr| self.to_addresses.contains(&to_addr));
+
+        enum Match {
+            All,
+            From,
+            To,
+            Both(TraceFilterMode),
         }
 
-        match self.mode {
-            TraceFilterMode::Union => {
-                self.from_addresses.contains(&from) ||
-                    to.map_or(false, |to| self.to_addresses.contains(&to))
-            }
-            TraceFilterMode::Intersection => {
-                self.from_addresses.contains(&from) &&
-                    to.map_or(false, |to| self.to_addresses.contains(&to))
-            }
+        let match_type = match (self.from_addresses.is_empty(), self.to_addresses.is_empty()) {
+            (true, true) => Match::All,
+            (false, true) => Match::From,
+            (true, false) => Match::To,
+            (false, false) => Match::Both(self.mode),
+        };
```

Comment:
this section can be moved to the top of the function since for (true, true) we don't need to check inclusion,

then we also don't really need the helper enum because all we're doing here is convert all 4 cases into an enum and match that again,

I'm okay with duplicating the inclusion checks in the From/To/Both arms here

Reply:
Pushed up that change, it does make it cleaner.